### PR TITLE
Add asymmetric JWT signing

### LINKED ISSUE #15912: Support RS256/JWKS for signing/verifying OAUTH JWTs
Background information on RS256 here: https://auth0.com/blog/navigating-rs256-and-jwks/

Utilizing RS256 (as an option) to sign JWTs means that a shared secret won't need to be shared with applications to verify the validity of the token (likely currently applications assume tokens are valid without checking signature).

Two applications that I tested using our OIDC well-known endpoint, which are Sourcegraph and Smallstep CA, fail due to them needing to verify tokens they receive.

I'm willing to payout a bounty of $100USD on this (minus whatever bogus fees paypal requires), and pay that directly to contributor who completes this ticket. This is instead of using bounty source as they takes slightly more off top than paypal directly (I'm going this way to incentivise completion even slightly more).

cc: @jonasfranz 

Edit: For this ticket please also create a jwks_uri and add it to the wellknown oidc endpoint.

### PR DESCRIPTION
**Pull request by KN4CK3R**
Close #15912

Added asymmetric JWT signing. Supports `HS256` (already implemented), `HS384`, `HS512`, `RS256`, `RS384`, `RS512`, `ES256`, `ES384` and `ES512`.

The default signing algorithm is changed from `HS256` (symmetric encryption) to `RS256` (asymmetric encryption).

## ⚠️  BREAKING ⚠️ 

* Asymmetric algorithms require a secret asymmetric key pair to be in `JWT_SIGNING_PRIVATE_KEY_FILE` (by default `APP_DATA_PATH/jwt`), a pair will be generated if it is not present. (NB: this was originally in `CUSTOM_PATH` but was changed by #16227)
* The original symmetric `JWT_SECRET` will only be used if `JWT_SIGNING_ALGORITHM` is set to `HS256` (previous default), `HS384` or `HS512`.
* As a result of the change of algorithm gitea OAuth2 tokens (and potentially client secrets) will need to be regenerated unless you change your `JWT_SIGNING_ALGORITHM` back to `HS256`.
* Legacy installations of Drone assume a short key length, however as no automated migrations are run by Drone you'll need to do them manually if the column type for user_oauth_token and user_oauth_refresh are limited to 500 characters:

```sql 
-- an example manual migration for Drone database connected to postgresql
alter table users alter column user_oauth_token type bytea using convert_to(user_oauth_token, 'LATIN1');
alter table users alter column user_oauth_refresh type bytea using convert_to(user_oauth_refresh, 'LATIN1');
```

or see: https://github.com/go-gitea/gitea/pull/16010#issuecomment-879381265

### CODE REVIEW

File: models/oauth2.go [hunk 1]
```diff
@@ -132,6 +132,9 @@ func GetActiveOAuth2Providers() ([]string, map[string]OAuth2Provider, error) {
 
 // InitOAuth2 initialize the OAuth2 lib and register all active OAuth2 providers in the library
 func InitOAuth2() error {
+	if err := oauth2.InitSigningKey(); err != nil {
```
Comment (lunny) [inline models/oauth2.go:135]: Merge it into oauth2.Init is better?

Comment (KN4CK3R) [reply models/oauth2.go:135]: I did not add it in the existing method because `oauth2.Init` is used for the other direction (Gitea -> other service). So I thought both methods should be logical separated.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 2]
```diff
@@ -350,7 +350,7 @@ relation to port exhaustion.
 - `ISSUE_INDEXER_PATH`: **indexers/issues.bleve**: Index file used for issue search; available when ISSUE_INDEXER_TYPE is bleve and elasticsearch.
 - The next 4 configuration values are deprecated and should be set in `queue.issue_indexer` however are kept for backwards compatibility:
 - `ISSUE_INDEXER_QUEUE_TYPE`: **levelqueue**: Issue indexer queue, currently supports:`channel`, `levelqueue`, `redis`.
-- `ISSUE_INDEXER_QUEUE_DIR`: **queues/common**: When `ISSUE_INDEXER_QUEUE_TYPE` is `levelqueue`, this will be the path where the queue will be saved. (Previously this was `indexers/issues.queue`.)
```
Comment (techknowlogick) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: per discussion in chat, these changed lines are fine due to the fact that it cleans up line endings (shakes fist at windows)

File: models/oauth2_application.go [hunk 3]
```diff
@@ -559,8 +559,8 @@ func ParseOAuth2Token(jwtToken string) (*OAuth2Token, error) {
 // SignToken signs the token with the JWT secret
 func (token *OAuth2Token) SignToken() (string, error) {
 	token.IssuedAt = time.Now().Unix()
-	jwtToken := jwt.NewWithClaims(jwt.SigningMethodHS512, token)
-	return jwtToken.SignedString(setting.OAuth2.JWTSecretBytes)
+	jwtToken := jwt.NewWithClaims(oauth2.DefaultSigningKey.SigningMethod(), token)
+	return jwtToken.SignedString(oauth2.DefaultSigningKey.SignKey())
```
Comment (techknowlogick) [inline models/oauth2_application.go:564]: Attempting to verify the generated token using JWT.io, it is unable to parse the signature. It works w/ HS256, but not RS256. I suspect it is something upstream in github.com/dgrijalva/jwt-go that is signing it differently than other libraries expect. I've tried to dig down into the code changes from this PR, but it looks like everything you are setting is correct.

Please let me know if you require any additional information from me to replicate.

Appendix:
You can also see at https://replit.com/@techknowlogick/BruisedCompetitiveLightweightprocess attempting to parse the key, that in headers it is unable to return KeyID (using smallstep CLI to fetch example token from google `step oauth --oidc --bare`, you can see that KeyID is populated).
Sample RS256 token:
```
eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiI5M2NiNDM5OS04Mzc4LTQ5NGYtYTQwNy1mYzVjODIzYTJjOTIiLCJleHAiOjE2MjI0NDUxMjcsImlhdCI6MTYyMjQ0MTUyNywiaXNzIjoiaHR0cHM6Ly90a2xrLWlubGV0cy5oZXJva3VhcHAuY29tLyIsInN1YiI6IjEiLCJub25jZSI6IjhmZWFlYmYxMzg0MzIwZDNmOTU4YTMxODE0MTQ1NWI5NDk2ZTg4NDA1NTNlZjk1NmVjODcyNGJiYTMzYzQ5MmIifQ.HfksMJPGgjhEP9UJqPulwPvh02VPTgmNkgAxGUdIn8PKiWbPcCkfT2yedFAMccduCqg3FQxLT_LEhPB0T8zuVl4wXdPS-t_qfGem1KE3sZubVdzy2PAA3Iya76kokbk5ufcfnt1bYpGp0p8sCGDm7oOma3N2yvsrXIWIC7EIzPzVICdD7a89ZFMC1TVw2zK_MKU9pUFYcuTrSfrb_Sx5OzpT3DxSMFJ0qfrVIdyYGlwbsujdxaBN2d1BcHv94Ur68FBxxbXoBqggIL7_gOZHmR8ENbRD1D8yJTTqk0pCjR42CqxaosEIv-vqFpLt92J_ZJCr4ldvA0EuVO9HhdjcaDPHrVVW4eh__QOI2aLLTuOcnM165LN5uGt5QN13Y9Ws6rU6inmhycPEqvdK11-o25Y3KlqwgKxNhJLyX-Ctr1Va1y5VZNNK6Oa30bF-_Mb4T92qELMoXthHi5d-wpiiwgURO2xfltxoBrfop0cA7-eCbQ2PPgYZ8oxkXCVGLtlwbcTQXEtzhpu7pz_YBptVtZtnGpqtQqXLcA_afDWuK7KE8xI3vgflSErFwx_QTPF8J_CPnf21IxXp5AoFNUcfZ-k0g7YxnpbX1tNctYGisieTIbW2Jp5iGODdAEKTzsb638jzrz3sOsE1ty01_IrpMkyfHbcbJYeA5SWiL1AA9l8
```

**Review by techknowlogick — CHANGES_REQUESTED** [review]
Thank you so much so far on your work on this PR!!

I was able to sign into Sourcegraph using this PR, however testing this PR more I found that 1. sourcegraph may not be using the cert to validate the signature on this token, 2. possibly the upstream jwt library may not be signing the token correctly.

File: models/oauth2_application.go (hunk 3 above)
Comment (KN4CK3R) [reply models/oauth2_application.go:564]: JWT.io works for me without changes. Did you change the "Algorithm" on top?
grafik

I used this container to get the token:
```
version: '3.5'

services:
  oidc-test-client:
    image: beryju/oidc-test-client
    ports:
      - 9009:9009
    network_mode: bridge
    environment:
      OIDC_CLIENT_ID: id
      OIDC_CLIENT_SECRET: secret
      OIDC_PROVIDER: gitea-endpoint
      OIDC_DO_INTROSPECTION: "false"
```
I then used
`openssl rsa -in private.pem -out rsa_priv.pem`
and
`openssl rsa -in rsa_priv.pem -pubout > rsa_pub.pem`
to get the private and public key in the desired format.

File: modules/auth/oauth2/jwtsigningkey.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+	return nil
+}
+
+// loadOrCreateSymmetricKey checks if the configured secret is valid.
+// If it is not valid a new secret is created and saved in the configuration file.
+func loadOrCreateSymmetricKey() (interface{}, error) {
+	key := make([]byte, 32)
+	n, err := base64.RawURLEncoding.Decode(key, []byte(setting.OAuth2.JWTSecretBase64))
+	if err != nil || n != 32 {
```
Comment (KN4CK3R) [inline modules/auth/oauth2/jwtsigningkey.go:299]: Currently only 32 bytes are allowed. Should we really enforce this? Smaller keys are valid and larger keys are hashed by the internal hashing algorithm to create a secret with the correct length.
https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/crypto/hmac/hmac.go#L148-L152

**Review by techknowlogick — APPROVED** [review]

File: modules/auth/oauth2/jwtsigningkey.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+type JWTSigningKey interface {
+	IsSymmetric() bool
+	SigningMethod() jwt.SigningMethod
+	SignKey() interface{}
+	VerifyKey() interface{}
+	ToJWK() (map[string]string, error)
+	PreProcessToken(*jwt.Token)
+}
+
+type hmacSingingKey struct {
```
Comment (jtran) [inline modules/auth/oauth2/jwtsigningkey.go]: This is the first instance, but there are lots of misspellings of "signing".

```suggestion
type hmacSigningKey struct {
```

File: modules/setting/setting.go [hunk 6]
```diff
@@ -801,21 +804,8 @@ func NewContext() {
...
-		n, err := base64.RawURLEncoding.Decode(OAuth2.JWTSecretBytes, []byte(OAuth2.JWTSecretBase64))
-
-		if err != nil || n != 32 {
-			OAuth2.JWTSecretBase64, err = generate.NewJwtSecret()
-			if err != nil {
-				log.Fatal("error generating JWT secret: %v", err)
-				return
-			}
-
-			CreateOrAppendToCustomConf(func(cfg *ini.File) {
```
Comment (6543) [inline modules/setting/setting.go:815]: should we realy drop this?

Comment (KN4CK3R) [reply modules/setting/setting.go:815]: It's not dropped. You can find it in `loadOrCreateSymmetricKey`

Comment (6543) [reply modules/setting/setting.go:815]: without this no `JWT_SECRET` will be created if there is none

Comment (6543) [reply modules/setting/setting.go:815]: so it's created on demand ... I would prefer if it get added on first gitea start ...

Comment (KN4CK3R) [reply modules/setting/setting.go:815]: It gets added on first gitea start if you use `HSxxx` as signing algorithm. If you use another algorithm it is not used at all.

File: modules/setting/setting.go [hunk 7]
```diff
@@ -371,14 +371,17 @@ var (
...
+		JWTSigningAlgorithm        string `ini:"JWT_SIGNING_ALGORITHM"`
 		JWTSecretBase64            string `ini:"JWT_SECRET"`
+		JWTSigningPrivateKeyFile   string `ini:"JWT_SIGNING_PRIVATE_KEY_FILE"`
 		MaxTokenLength             int
 	}{
 		Enable:                     true,
 		AccessTokenExpirationTime:  3600,
 		RefreshTokenExpirationTime: 730,
 		InvalidateRefreshTokens:    false,
+		JWTSigningAlgorithm:        "RS256",
```
Comment (6543) [inline modules/setting/setting.go:383]: this change default alg to async - not sure if it is breaking ...

Comment (6543) [reply modules/setting/setting.go:383]:
```suggestion
		JWTSigningAlgorithm:        "HS256",
```

Comment (6543) [reply modules/setting/setting.go:383]: as per KN4CK3R ...
`https://accounts.google.com/.well-known/openid-configuration` and other big tech have **RS256** - so go for it

Comment (KN4CK3R) [reply modules/setting/setting.go:383]: For clarification:
RS256 is defined as "create the signature with SHA**256** and use the provided RSA key as secret". The RSA key is allowed to have any size and **not** just 256 bit. If Gitea generates a key the default size is 4096 bit.

**Review by 6543 — APPROVED** [review]

**Comment by proton-ab** [comment]
Friendly warning to anyone out there using Drone - it fails to support new tokens with are too long
```"msg": "cannot update user: pq: value too long for type character varying(500)",```

**Comment by markeissler** [comment]
**TL;DR:** For Drone compatibility... Set the gitea environment variable to:

```
GITEA__OAUTH2__JWT_SIGNING_ALGORITHM=HS256
```

And restart gitea - OR - alter your database to accommodate the longer value that needs to be stored. Something like the following might work, or perhaps set the column type to TEXT instead of BYTEA.

```sql
ALTER TABLE users ALTER COLUMN user_oauth_token BYTEA;
```

I would suggest modifying the column in the drone db for production environments since the intention is to improve security with asymmetric tokens. By default, the column is setup as a `varchar(500)` type.

**Comment by KN4CK3R** [comment]
A shorter RSA key may help too. The default key size is 4096 bit.
The drone problem was fixed in this commit two months ago:
https://github.com/drone/drone/commit/eaadd82a839b89ddcc9e0455132fc86ff6f827d6

**Comment by 6543** [comment]
thanks for the hint, for more issues etc please open a new issues
